{# #######################
CpG-strandness.html
##########################

diagnosis of CpG strandness of sequencing coverage and DNAme level.

#}

<section id="CpG-strandness">
    <h2>CpG strandness</h2>
    <p>Summarise the reads mapped onto double strands of CpGs. </p>

    <article id="CpG-stranded-depth">
        <h3>CpG read depth of two strands</h3>
        <p>
            The heatmap shows the number of CpGs (instead if cytosines) with
            different Watson (row) and Crick (column) read depths. Uncovered
            CpGs with coordinate (0,0) is removed for better visualization.
        </p>
        <img src="img/heatmap-cg-stranded-depth.png" id="fig-heatmap-cg-stranded-depth" alt="" width="500px" />
        <div class="caption" link="#fig-heatmap-cg-stranded-depth">This picture is useful to diagnose the stranded pattern of whole-genome CpG coverage.</div>
    </article>

    <article id="CpG-depth-diff">
        <h3>Depth difference between strands</h3>
        <p>
        <ul>
            <li>CpG depth difference = Watson depth - Crick depth </li>
            <li>difference: mean = {{alldata.mean_strand_dp_diff}}, sd<sup class="question-marker" id="cap-strandness-depth-diff">?</sup> = {{alldata.sd_strand_dp_diff}} </li>
            <div class="caption" link="#cap-strandness-depth-diff">
                sd = standard deviation
            </div>
            <li>absolute difference: mean = {{alldata.mean_abs_strand_dp_diff}}, sd = {{alldata.sd_abs_strand_dp_diff}} </li>
        </ul>
        </p>
        <img src="img/cg-stranded-depth-difference.png" alt="" height="250px" />
    </article>

    <article id="CpG-double-stranded-coverage">
        <h3>CpGs of double-stranded coverage</h3>
        <p>
            Proportion of double-stranded covered CpGs at different depths.
            $$ \text{proportion of CpGs of double-stranded coverage} = \frac{\#\{
            \text{CpGs with Watson and Crick depths both} \ge \text{DP}\}}
            {\#\{ \text{CpGs with either Watson or Crick depths} \ge
            \text{DP}\}} $$
        </p>
        <img src="img/cg-prop-of-double-stranded-coverage.png" alt="" height="300px" />
    </article>

    <article id="CpG-stranded-meth">
        <h3>CpG DNAme level of two strands</h3>
        <p>
            The following heatmap shows the numbers of sampled CpGs of different stranded methylation. Read depth â‰¥ 1 on both strands.
        </p>
        <img src="img/heatmap-cg-stranded-meth.png" alt="" width="500px" />
    </article>
    
    <article id="CpG-stranded-meth-diff">
        <h3>DNAme level difference between strands</h3>
        <p>
            The following heatmap shows the number of CpGs with different
            methylation difference (Watson - Crick, y-axis) at different depth
            (x-axis), represented by $\min\{ \text{depth}_\text{Watson},
            \text{depth}_\text{Crick} \} $.
        </p>
        <img src="img/heatmap-cg-stranded-meth-diff.png" alt="" height="400px" />
    </article>
</section>

