{# #######################
plastid.html
##########################

diagnosis of reads mapped to plastid DNA.

#}

<section id="plastid">
    <h2>Plastid</h2>

    <article id="plastid-coverage">
        <ul>
            <li>{{ alldata.plastid_length }} bps in total, whereas {{ alldata.plastid_cov_prop }} are covered ({{ alldata.plastid_covn }} bp for depth ≥ 1);</li>
            <li>{{ alldata.plastid_nCorG }} cytosines in total, whereas {{ alldata.plastid_covnCorG_prop }} are covered ({{ alldata.plastid_covnCorG }} cytosines for depth ≥ 1).</li>
        </ul>
        <h3>Read depth</h3>
        <p>
            Mean read depth of {{ alldata.plastid_bin_size }}bp binning windows for Watson strand (top), Crick strand (middle), and double strands (bottom).
        </p>
        <img src="img/depth-bin-plastid.png" alt="" width="300px" id="img-depth-bin-plastid" />
        <div class="caption" link="#img-depth-bin-plastid">
            Check the uniformness of read coverage.
        </div>
    </article>

    <article id="plastid-dna-copy-number">
        <h3>DNA copy number</h3>
        <p>
            Plastid DNA copy number is estimated by the average depth of plastids over
            that of nuclear chromosomes. <sup id="plastid-dna-copy-number" class="question-marker">?</sup>
        <div class="caption" link="#plastid-dna-copy-number">
            The estimate is less accurate for RRBS samples than WGBS samples due to selective enrichment of DNA fragments. <br> Change the coefficient accordingly if not diploidy.
        </div>
        {% raw %} $$ \text{plastid DNA copy number} = \frac{ {% endraw %} {{ alldata.plastid_mean_dp}} {% raw %} }{ {% endraw %}
        {{ alldata.mean_dp }} {% raw %} } \times 2 (\text{diploidy}) = {% endraw %} {{ alldata.plastid_copynum }} $$
        </p>
    </article>

    <article id="plastid-conversion">
        <h3>Bisulfite conversion</h3>
        <p>
            From the 'methylation' of cytosines in plastid genome
            <sup id="plastid-conversion" class="question-marker">?</sup>, we can estimate bisulfite conversion rate as 
            $$ \text{conversion rate} = 1-\text{plastid methylation level} = 1 - {{ alldata.plastid_me }} = {{ alldata.bsrate_plastid }}\%. $$
        </p>
        <div class="caption" link="#plastid-conversion">
            Assume plastid DNA is not 5mC methylated.
        </div>
    </article>

    <article id="plastid-error">
        <h3>Error rate</h3>
        <p>
            The (base-level) error rate that a base is mis-detected during library
            preparation, sequencing, and alignment, can be estimated
            <sup id="plastid-error" class="question-marker">?</sup> from inconsistent
            bases mapped to plastid genome. From the error rate distribution of
            20-bp binning windows, the average error rate is 
            $$ \text{error rate} =
            \frac{\text{bases different with dominating base}}{\text{total read
            bases}} = {{ alldata.err_rate_plastid}} \%. $$
        </p>
        <div class="caption" link="#plastid-error">
            Mutation rate is much smaller than error rate in bisulfite-seq samples.
        </div>
        <img src="img/base-error-rate-dist-plastid.png" alt="" width="400px" />
    </article>
</section>
