{# #######################
summary.html
##########################

summary information.

#}

<section class="first-section" id="Summary">
    <h2 class="first-header">Summary</h2>
    <table class="table table-striped table-sm table-hover">
        <!-- <thead>
            <tr>
                <th scope="col">#</th>
                <th scope="col">First</th>
                <th scope="col">Last</th>
                <th scope="col">Handle</th>
            </tr>
        </thead> -->
        <tbody>
            <tr>
                <td class="table-header">Coverage and DNAme level</td>
                <td>depth ≥ 1</td> <td>depth ≥ 3</td> <td>depth ≥ 5</td> <td>depth ≥ 10</td>
            </tr>
            <tr>
                <td colspan="5" class="tb-major">Whole-genome Coverage: </td>
            </tr>
            <tr>
                <td class="tb-minor">single strand</td>
                <td>{{ alldata.covrate_wg_sp[0] }}%</td>
                <td>{{ alldata.covrate_wg_sp[1] }}%</td>
                <td>{{ alldata.covrate_wg_sp[2] }}%</td>
                <td>{{ alldata.covrate_wg_sp[3] }}%</td>
            </tr>
            <tr>
                <td class="tb-minor">two strands</td>
                <td>{{ alldata.covrate_wg_dp[0] }}%</td>
                <td>{{ alldata.covrate_wg_dp[1] }}%</td>
                <td>{{ alldata.covrate_wg_dp[2] }}%</td>
                <td>{{ alldata.covrate_wg_dp[3] }}%</td>
            </tr>

            <tr>
                <td colspan="5" class="tb-major">Cytosine Coverage:</td>
            </tr>
            <tr>
                <td class="tb-minor">CG</td>
                <td>{{ alldata.covrate_cg[0] }}%</td>
                <td>{{ alldata.covrate_cg[1] }}%</td>
                <td>{{ alldata.covrate_cg[2] }}%</td>
                <td>{{ alldata.covrate_cg[3] }}%</td>
            </tr>
            <tr>
                <td class="tb-minor">CHG</td>
                <td>{{ alldata.covrate_chg[0] }}%</td>
                <td>{{ alldata.covrate_chg[1] }}%</td>
                <td>{{ alldata.covrate_chg[2] }}%</td>
                <td>{{ alldata.covrate_chg[3] }}%</td>
            </tr>
            <tr>
                <td class="tb-minor">CHH</td>
                <td>{{ alldata.covrate_chh[0] }}%</td>
                <td>{{ alldata.covrate_chh[1] }}%</td>
                <td>{{ alldata.covrate_chh[2] }}%</td>
                <td>{{ alldata.covrate_chh[3] }}%</td>
            </tr>
            <tr>
                <td class="tb-major">CpGs of double-stranded coverage <sup class="question-marker" id="cap-summary-double-cov">?</sup>
                <div class="caption" link="#cap-summary-double-cov">
                    #{CpGs with depth ≥ k on both strands} over #{CpGs with depth ≥ k on any strand}
                </div></td>
                <td>{{ alldata.covrate_cg_ds[0] }}%</td>
                <td>{{ alldata.covrate_cg_ds[1] }}%</td>
                <td>{{ alldata.covrate_cg_ds[2] }}%</td>
                <td>{{ alldata.covrate_cg_ds[3] }}%</td>
            </tr>

            <tr>
                <td colspan="5" class="tb-major">DNAme level: </td>
            </tr>
            <tr>
                <td class="tb-minor">CG</td>
                <td>{{ alldata.me_cg_ds[0] }}%</td>
                <td>{{ alldata.me_cg_ds[1] }}%</td>
                <td>{{ alldata.me_cg_ds[2] }}%</td>
                <td>{{ alldata.me_cg_ds[3] }}%</td>
            </tr>
            <tr>
                <td class="tb-minor">CHG</td>
                <td>{{ alldata.me_chg_ds[0] }}%</td>
                <td>{{ alldata.me_chg_ds[1] }}%</td>
                <td>{{ alldata.me_chg_ds[2] }}%</td>
                <td>{{ alldata.me_chg_ds[3] }}%</td>
            </tr>
            <tr>
                <td class="tb-minor">CHH</td>
                <td>{{ alldata.me_chh_ds[0] }}%</td>
                <td>{{ alldata.me_chh_ds[1] }}%</td>
                <td>{{ alldata.me_chh_ds[2] }}%</td>
                <td>{{ alldata.me_chh_ds[3] }}%</td>
            </tr>
        </tbody>
    </table>

    <table class="table table-striped table-sm table-hover">
        <tbody>
            <tr>
                <td class="table-header"> Reads Composition</td>
                <td>Nuclear DNA</td>
                <td>MT DNA</td>
                <td>Plastid DNA</td>
                <td>Control DNA</td>
            </tr>
            <tr>
                <td>Reads<br>(%)</td>
                <td>{{alldata.nreads.nuclear}}<br>({{alldata.nreads_prop.nuclear}}%)</td>
                <td>{{alldata.nreads.mt}}<br>({{alldata.nreads_prop.mt}}%)</td>
                <td>{{alldata.nreads.plastid}}<br>({{alldata.nreads_prop.plastid}}%)</td>
                <td>{{alldata.nreads.lambda}}<br>({{alldata.nreads_prop.lambda}}%)</td>
            </tr>
            <tr>
                <td>Mean read depth</td>
                <td>{{alldata.mean_dp}}</td>
                <td> 
                {% if alldata.mt_is_covered == 1 %}
                    {{alldata.mt_mean_dp}}
                {% else %}
                    -
                {% endif %} </td>
                <td> 
                {% if alldata.plastid_is_covered == 1 %}
                    {{alldata.plastid_mean_dp}}
                {% else %}
                    -
                {% endif %} </td>
                <td> 
                {% if alldata.lambda_is_covered == 1 %}
                    {{alldata.lambda_mean_dp}}
                {% else %}
                    -
                {% endif %} </td>
            </tr>
            <!-- <tr>
                <td>Mean read depth</td>
                <td>{{alldata.mean_dp_dict.nuclear}}</td>
                <td>{{alldata.mean_dp_dict.mt}}</td>
                <td>{{alldata.mean_dp_dict.plastid}}</td>
                <td>{{alldata.mean_dp_dict.lambda}}</td>
            </tr> -->
            <tr>
                <td>DNA copy number</td>
                <td>{{alldata.ploidy}} (ploidy)</td>
                <td> {% if alldata.mt_is_covered == 1 %}
                    {{alldata.mt_copynum}}
                    {% else %}
                    -
                    {% endif %} </td>
                <td> {% if alldata.plastid_is_covered == 1 %}
                    {{alldata.plastid_copynum}}
                    {% else %}
                    -
                    {% endif %} </td>
                <td> {% if alldata.lambda_is_covered == 1 %}
                    {{alldata.lambda_copynum}}
                    {% else %}
                    -
                    {% endif %} </td>
            </tr>
            <!-- <tr>
                <td>DNA copy number</td>
                <td>{{alldata.copy_num.nuclear}} (ploidy)</td>
                <td>{{alldata.copy_num.mt}}</td>
                <td>{{alldata.copy_num.plastid}}</td>
                <td>{{alldata.copy_num.lambda}}</td>
            </tr> -->

            <tr><td colspan="5" class="table-header"> Conversion and Error</td></tr>

            <tr>
                <td>Bisulfite Conversion Rate <sup class="question-marker" id="cap-summary-coversion-rate">?</sup>
                    <div class="caption" link="#cap-summary-coversion-rate">
                        Assuming nuclear CHHs and cytosines of extranuclear DNA are not methylated.
                    </div>
                </td>
                <td>{{alldata.bsrate_chh}}%</td>
                <td> {% if alldata.mt_is_covered == 1 %}
                    {{alldata.bsrate_mt}}%
                    {% else %} -
                    {% endif %} </td>
                <td> {% if alldata.plastid_is_covered == 1 %}
                    {{alldata.bsrate_plastid}}%
                    {% else %} -
                    {% endif %} </td>
                <td> {% if alldata.lambda_is_covered == 1 %}
                    {{alldata.bsrate_lambda}}%
                    {% else %} -
                    {% endif %} </td>
            </tr> 
            <tr>
                <td>Base-level Error Rate</td>
                <td>{{alldata.err_rate_AT}}%</td>
                <td> {% if alldata.mt_is_covered == 1 %}
                    {{alldata.err_rate_mt}}%
                    {% else %} -
                    {% endif %} </td>
                <td> {% if alldata.plastid_is_covered == 1 %}
                    {{alldata.err_rate_plastid}}%
                    {% else %} -
                    {% endif %} </td>
                <td> {% if alldata.lambda_is_covered == 1 %}
                    {{alldata.err_rate_lambda}}%
                    {% else %} -
                    {% endif %} </td>
            </tr>
        </tbody>
    </table>
    <table class="table table-striped table-sm table-hover">
        <tbody>
        <tr><td colspan="4" class="table-header">DNA Loss and Sequencing Saturation</td></tr>
            <tr>
                <td>Unsequenced Nuclear DNA: {{alldata.prop_unsequenced}}% </td>
                <td>Sequencing Saturation: {{alldata.saturation}}% </td>
                <td>Lost Nuclear DNA: {{alldata.prop_lost}}% </td>
                <td>Max Coverage Rate: {{alldata.prop_asymptotic}}% <sup class="question-marker" id="cap-summary-max-coverage-rate">?</sup>
                    <div class="caption" link="#cap-summary-max-coverage-rate">
                        Whole-genome coverage rate by sequencing continuously. 
                    </div></td>
            </tr>
        </tbody>
    </table>

    <article>
        <h3>Reads mapped to each chr/contig</h3>
        <div class="img" id="plot-reads-vs-lengths" style="width: 450px; height: 350px"></div>
    </article>

    <article>
        <h3>Read/DNA composition</h3>
        DNA content <sup id="cap-DNA-composition" class="question-marker">?</sup> (measured by mol of nucleotides ) ∝ #reads, assuming reads are of the same length.
        <div class="caption" link="#cap-DNA-composition">
            DNA content in the sequencing library.
        </div>

        <img src="img/pie-DNA-content.png" alt="" height="250px" />
    </article>
</section>
