{# #######################
summary.html
##########################

summary information.

#}

<section class="first-section" id="Summary">
    <h2 class="first-header">Summary</h2>
    <table class="table table-striped table-sm">
        <caption>List of users</caption>
        <!-- <thead>
            <tr>
                <th scope="col">#</th>
                <th scope="col">First</th>
                <th scope="col">Last</th>
                <th scope="col">Handle</th>
            </tr>
        </thead> -->
        <tbody>
            <tr>
                <td></td>
                <td>depth ≥ 1</td> <td>depth ≥ 3</td> <td>depth ≥ 5</td> <td>depth ≥ 10</td>
            </tr>
            <tr>
                <td colspan="5" class="tb-major">Whole-genome Coverage: </td>
            </tr>
            <tr>
                <td class="tb-minor">double strand:</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
            </tr>
            <tr>
                <td class="tb-minor">Watson strand:</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
            </tr>
            <tr>
                <td class="tb-minor">Crick strand:</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
            </tr>

            <tr>
                <td colspan="5" class="tb-major">CpG Coverage: </td>
            </tr>
            <tr>
                <td class="tb-minor">double strand:</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
            </tr>
            <tr>
                <td class="tb-minor">Watson strand:</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
            </tr>
            <tr>
                <td class="tb-minor">Crick strand:</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
            </tr>

            <tr>
                <td colspan="5" class="tb-major">DNAme level: </td>
            </tr>
            <tr>
                <td class="tb-minor">CG:</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
            </tr>
            <tr>
                <td class="tb-minor">CHG:</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
            </tr>
            <tr>
                <td class="tb-minor">CHH:</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
                <td>{{ alldata.bsrate_chh }}%</td>
            </tr>
        </tbody>
    </table>

    <table class="table table-striped table-sm">
        <tbody>
            <tr>
                <td>Read Depth</td><td>10</td>
            </tr>
        </tbody>
    </table>

    <table class="table table-striped table-sm">
        <tbody>
            <tr>
                <td></td>
                <td>From CHHs</td>
                <td>From MT DNA</td>
                <td>From plastid DNA</td>
                <td>From lambda DNA/spiked-in DNA</td>
            </tr>
            <tr>
                <td>Bisulfite Conversion Rate:</td>
                <td></td><td></td><td></td><td></td>
            </tr>
            <tr>
                <td>Base-level Error Rate:</td>
                <td></td>
                <td></td>
                <td></td>
                <td></td>
            </tr>
        </tbody>
    </table>
    <table class="table table-striped table-sm">
        <tbody>
            <tr>
                <td>Sequencing Saturation:</td>
                <td>DNA Template Loss</td>
            </tr>
        </tbody>
    </table>

    <article>
        <h3>Reads mapped to each chr/contig</h3>
        <div class="img" id="plot-reads-vs-lengths" style="width: 450px; height: 350px"></div>
    </article>

    <article>
        <h3>DNA composition</h3>
        According to reads mapped to nuclear and extranuclear chromosomes/contigs, DNA content in the sequencing library is estimated.

        <img src="img/pie-DNA-content.png" alt="" height="250px" />
    </article>
</section>
